corehq/apps/couch_sql_migration/couchsqlmigration.py

```diff
@@ -175,60 +181,71 @@ def _migrate_unprocessed_form(self, couch_form_json):
+            self.log_info('Waiting on {} docs'.format(len(pool)))
+
+    def _copy_unprocessed_case(self, change):
+        couch_case = CommCareCase.wrap(change.get_document())
+        self.log_debug('Processing doc: {}({})'.format(couch_case['doc_type'], change.id))
+        try:
+            first_action = couch_case.actions[0]
+        except IndexError:
+            first_action = CommCareCaseAction()
+
+        sql_case = CommCareCaseSQL(
+            case_id=couch_case.case_id,
+            domain=self.domain,
+            type=couch_case.type or '',
+            name=couch_case.name,
+            owner_id=couch_case.owner_id or couch_case.user_id or '',
+            opened_on=couch_case.opened_on or first_action.date,
+            opened_by=couch_case.opened_by or first_action.user_id,
+            modified_on=couch_case.modified_on,
+            modified_by=couch_case.modified_by or couch_case.user_id or '',
+            server_modified_on=couch_case.server_modified_on,
+            closed=couch_case.closed,
+            closed_on=couch_case.closed_on,
+            closed_by=couch_case.closed_by,
+            deleted=True,
+            deletion_id=couch_case.deletion_id,
+            deleted_on=couch_case.deletion_date,
+            external_id=couch_case.external_id,
+            case_json=couch_case.dynamic_case_properties()
+        )
+        _migrate_case_actions(couch_case, sql_case)
+        _migrate_case_indices(couch_case, sql_case)
+        _migrate_case_attachments(couch_case, sql_case)
+        try:
+            CaseAccessorSQL.save_case(sql_case)
+        except IntegrityError:
+            # case re-created by form processing so just mark the case as deleted
+            CaseAccessorSQL.soft_delete_cases(
+                self.domain,
+                [sql_case.case_id],
+                sql_case.deleted_on,
+                sql_case.deletion_id
             )
-            _migrate_case_actions(couch_case, sql_case)
-            _migrate_case_indices(couch_case, sql_case)
-            _migrate_case_attachments(couch_case, sql_case)
-            try:
-                CaseAccessorSQL.save_case(sql_case)
-            except IntegrityError:
-                # case re-created by form processing so just mark the case as deleted
-                CaseAccessorSQL.soft_delete_cases(
-                    self.domain,
-                    [sql_case.case_id],
-                    sql_case.deleted_on,
-                    sql_case.deletion_id
-                )
 
     def _calculate_case_diffs(self):
         cases = {}
+        pool = Pool(5)
         changes = _get_case_iterator(self.domain).iter_all_changes()
         for change in self._with_progress(CASE_DOC_TYPES, changes, progress_name='Calculating diffs'):
             cases[change.id] = change.get_document()
             if len(cases) == 1000:
-                self._diff_cases(cases)
+                pool.spawn(self._diff_cases, deepcopy(cases))
```

Comment:
`deepcopy` can be a relatively expensive operation, depending on the complexity of the objects being copied. This may be an area to consider for optimization if it turns out to be a bottleneck.

Reply:
It doesn't seem to be so far, but I'll keep it in mind